### PR TITLE
document Hegg Dongle

### PR DESCRIPTION
**Pull request by Habbie**
~~this will need a bit more work, but PRing it now allows me to share the rendered preview with my collaborators~~

done!

### CODE REVIEW

**Comment by vandalon** [comment]
Sample yaml config:

```
esphome:
  name: "p1-reader"
  on_boot:
    then:
      - light.turn_on:
         id: status_led
         brightness: 50%
         blue: 0
         red: 0
         green: 100%
         effect: pulse

esp8266:
  board: esp12e

ota:
  password: !secret api_pass

api:
  encryption:
    key: !secret api_key
  on_client_connected:
      - light.turn_on:
         id: status_led
         brightness: 50%
         blue: 0.949
         red: 0.094
         green: 0.737
         effect:  none
  on_client_disconnected:
      - light.turn_on:
         id: status_led
         brightness: 50%
         blue: 0.949
         red: 0.094
         green: 0.737
         effect:  pulse

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  fast_connect: true
  power_save_mode: HIGH
  on_connect:
      - light.turn_on:
         id: status_led
         brightness: 50%
         blue: 0
         red: 0
         green: 100%
         effect:  none
  on_disconnect:
      - light.turn_on:
         id: status_led
         brightness: 50%
         blue: 0
         red: 100%
         green: 0
         effect: pulse

logger:
  baud_rate: 0

captive_portal:

web_server:
  port: 80

uart:
  rx_pin: 3
  tx_pin: 1
  baud_rate: 115200
  rx_buffer_size: 1700

light:
  - platform: rgb
    internal: True
    id: status_led
    name: "Status Led"
    red: red
    green: green
    blue: blue
    effects:
      - pulse:

sensor:
  - platform: dsmr
    energy_delivered_tariff1:
      name: "Energy Consumed Tariff 1"
      state_class: total_increasing
    energy_delivered_tariff2:
      name: "Energy Consumed Tariff 2"
      state_class: total_increasing
    energy_returned_tariff1:
      name: "Energy Produced Tariff 1"
      state_class: total_increasing
    energy_returned_tariff2:
      name: "Energy Produced Tariff 2"
      state_class: total_increasing
    power_delivered:
      name: "Power Consumed"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_returned:
      name: "Power Produced"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    electricity_failures:
      name: "Electricity Failures"
      icon: mdi:alert
    electricity_long_failures:
      name: "Long Electricity Failures"
      icon: mdi:alert
    voltage_l1:
      name: "Voltage Phase 1"
    voltage_l2:
      name: "Voltage Phase 2"
    voltage_l3:
      name: "Voltage Phase 3"
    current_l1:
      name: "Current Phase 1"
    current_l2:
      name: "Current Phase 2"
    current_l3:
      name: "Current Phase 3"
    power_delivered_l1:
      name: "Power Consumed Phase 1"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_delivered_l2:
      name: "Power Consumed Phase 2"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_delivered_l3:
      name: "Power Consumed Phase 3"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_returned_l1:
      name: "Power Produced Phase 1"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_returned_l2:
      name: "Power Produced Phase 2"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    power_returned_l3:
      name: "Power Produced Phase 3"
      unit_of_measurement: "W"
      state_class: "measurement"
      accuracy_decimals: 0
      filters:
        - multiply: 1000
    gas_delivered:
      name: "Gas Consumed"
      state_class: total_increasing

text_sensor:
  - platform: dsmr
    identification:
      name: "DSMR Identification"
    p1_version:
      name: "DSMR Version"

output:
  - platform: esp8266_pwm
    pin: 14
    frequency: 600 Hz
    id: blue
  - platform: esp8266_pwm
    pin: 13
    frequency: 600 Hz
    id: red
  - platform: esp8266_pwm
    pin: 12
    frequency: 600 Hz
    id: green
```

**Comment by bitkat** [comment]
Some small nits for the following text:
_The DSMR/P1 smart meter protocol uses inverted voltages for serial, with the receiver (the dongle in your hand right now) acting as an Open Collector. This means that with no smart meter configured, RXD0 is not idle._

Firstly, the dongle doesn't "act as" an open collector; the P1-output *is* the open collector. It can pull the dongle's input to ground.

Secondly, I think you mean to say that RX isn't idle when the dongle is not *connected* to a smart meter. No configuration involved...

File: src/docs/devices/Hegg-Dongle/index.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+| ------ | ---------------------------- |
+| GND    | many; connected to one side of the (unpopulated) SW1 header; also available as the unlabeled pad |
+| GPIO0  | FLASH, connected to the other side of the (unpopulated) SW1 header
+| GPIO1  | TXD0, "txd" pad |
+| GPIO3  | RXD0, "rxd" pad, "data_in" pad via inverter |
+| GPIO12 | green 
+| GPIO13 | red
+| GPIO14 | blue
+| VCC    | power, connected to "3.3v" pad
+| ???    | P1 data request?
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
| Pin    | Function                     |Pad         |
| ------ | ---------------------------- | --------------------------- |
| GND    | many; connected to one side of the (unpopulated) SW1 header; also available as the unlabeled pad | Unlabeled |
| GPIO0  | FLASH, connected to the other side of the (unpopulated) SW1 header||
| GPIO1  | TXD0| txd |
| GPIO3  | RXD0 | RXD or data_in via inverter|
| GPIO12 | green|| 
| GPIO13 | red||
| GPIO14 | blue||
| VCC    | power|3.3v|
| ???    | P1 data request?||
```

File: src/docs/devices/Hegg-Dongle/index.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+## Flashing procedure
+
+Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
+The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
+It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.
+To make RXD0 idle, it is enough to tie P1 Data to Ground.
+To do that, you can either use the pads (`data_in` and FIXME is there a GND pad? perhaps next to 3.3v?), or tie them together on the P1 connector end.
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.

To flash:
* Make RXD0 idle by connecting P1 Data to Ground, or tie the corresponding pins on the P1 connector end together.
* Connect either the txd pad or the TXD0 pin to RX on your serial converter
* Connect either the rxd or data_in pad or the RXD0 pin to TX on your serial converter
* Power by connecting GND and Power to your serial converter
* Flash as normal (FIXME: do we need to add speeds / serial settings?)

```

File: src/docs/devices/Hegg-Dongle/index.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+A (so far) untested alternative is to poke a few pins through the plastic at the bottom to reach the pads.
+If you try this, please update this document with your experiences!
+
+Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
+The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
+It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.
+
+To flash:
+
+* Make RXD0 idle by connecting `data_in` to Ground, or tie the corresponding pins (5 and 6, the topmost two) on the P1 connector end together.
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
* Make GPIO3/RXD0 idle by connecting `data_in` to GND, or tie the corresponding pins (5 and 6, the topmost two) on the P1 connector end together.
```

File: src/docs/devices/Hegg-Dongle/index.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+* Connect either the `txd` pad or the TXD0 pin to RX on your serial converter
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
* Connect either the `txd` pad or the GPIO1/TXD0 pin to RX on your serial converter
```

File: src/docs/devices/Hegg-Dongle/index.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+* Connect either the `rxd`, or `data_in` pad, or pin 5 of the P1 socket, or the RXD0 pin to TX on your serial converter. If you use `data_in` or the P1 pin, you likely need to invert your signal.
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
* Connect either the `rxd`, or `data_in` pad, or pin 5 of the P1 socket, or the GPIO3/RXD0 pin to TX on your serial converter. If you use `data_in` or the P1 pin, you likely need to invert your signal.
```

File: src/docs/devices/Hegg-Dongle/index.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+* Power by connecting GND and Power to your serial converter, or plugging in USB. Presumably you can also power the dongle via 5V+GND on the P1 connector. Have GND connected to GPIO0 during powerup, or briefly tie RST to GND with GPIO0 also tied to GND to reset into flashing mode.
```
Comment (rvdm) [inline src/docs/devices/Hegg-Dongle/index.md]:
```suggestion
* Power by connecting GND and VCC to your serial converter, or plugging in USB. Presumably you can also power the dongle via 5V+GND on the P1 connector. Have GND connected to GPIO0/FLASH during powerup, or briefly tie RST to GND with GPIO0/FLASH also tied to GND to reset into flashing mode.
```

**Review by rvdm — COMMENTED** [review]
Added consistent PIN+meaning naming in instructions.

**Review by Habbie — COMMENTED** [review]
consistency, thanks @rvdm

**Comment by Habbie** [comment]
rebased and squash, ready to go I hope!

**Pull request merged**